addons/product/models/product_template.py

```diff
@@ -938,6 +939,83 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            return
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # The following dict uses product template attribute values as keys
+        # 0 means the value is acceptable, greater than 0 means it's rejected, it cannot be negative
+        # Bear in mind that several values can reject the same value and the latter can only be included in the
+        #  considered combination if no value rejects it.
+        # This dictionary counts how many times each value is rejected.
+        # Each time a value is included in the considered combination, the values it rejects are incremented
+        # When a value is discarded from the considered combination, the values it rejects are decremented
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values)
+        # determines which attribute line we're working on
+        attribute_idx = 0
+
+        def current_value_id():
+            return product_template_attribute_values[attribute_idx][values_idx[attribute_idx]].id
+
+        while True:
+            if values_idx[attribute_idx] >= 0:
+                for value_to_include_back in all_exclusions[current_value_id()]:
+                    current_exclusions[value_to_include_back] -= 1
+                included_values_ids.discard(current_value_id())
+
+            # first we try to get to next value of current attribute
+            if values_idx[attribute_idx] < len(product_template_attribute_values[attribute_idx]) - 1:
+                # get to next value of current attribute line
+                values_idx[attribute_idx] += 1
+
+                if current_exclusions[current_value_id()] or \
+                        any(value_to_exclude in included_values_ids for value_to_exclude
+                            in all_exclusions[current_value_id()]):
+                    continue
```

Comment:
add comment (or move the 2 parts of the condition into accordingly named variables):
"don't include the current value if it is already excluded by another included value, or if it would exclude an already included value"

Reply:
done